### PR TITLE
Refactor codec traits.

### LINKED ISSUE #24: Removing the Hash trait
Would it make sense to remove the `Hash` trait?

https://github.com/ipfs-rust/rust-ipld/blob/e172fbc8b7c83c751a82c0d4dde877885335152f/base/src/hash.rs#L7-L14

With the new rust-multihash version, it is just the same as `MultihashDigest`:

https://github.com/multiformats/rust-multihash/blob/a7996e06782a703fb983ed665564add0bbf89d51/src/digests.rs#L275-L285

so we could use that directly.

### LINKED ISSUE #27: Replace WriteCbor and ReadCbor.
Instead of `WriteCbor` and `ReadCbor` we should use something like `Encode<Codec::Cbor>` and `Decode<Codec::Cbor>`. On the derive side we could potentially add a generic derive that would be used for `dag-pb` and `dag-json` via an intermediate encoding to ipld and still keep the custom code for cbor that is much faster.

### LINKED ISSUE #34: add raw codec

### PR DESCRIPTION
**Pull request by dvc94ch**
Refactors the codec trait and updates libipld to use the new store and codec traits. The old store/gc file have been removed.

Depends on https://github.com/multiformats/rust-cid/pull/52 and https://github.com/multiformats/rust-multihash/pull/65

closes #24 closes #27 closes #34 

Created a follow-up issue #35

Moved the block builder to https://github.com/dvc94ch/ipld-block-builder

### CODE REVIEW

File: core/src/hash.rs
Comment (dvc94ch) [inline core/src/hash.rs:1]: TODO: move this to multihash

File: core/src/mem.rs
Comment (dvc94ch) [inline core/src/mem.rs]: this is actually wrong

File: Cargo.toml
Comment (aphelionz) [inline Cargo.toml]: note: we'll need to remove these before merging

File: core/src/mem.rs
Comment (aphelionz) [inline core/src/mem.rs]: As I understand this implementation is a few things - reference implementation for a Store, something built-in that rust-ipld can use if nothing else is implementing a store from the outside... is that correct?

Wondering if this would be better considered as a testing util and/or something in the examples folder? But maybe I'm missing the purpose of this.

Comment (aphelionz) [reply core/src/mem.rs]: The above is mostly just a folder structure nitpick, but I think nailing that down will help future implementers / readers if it's categorized properly.

File: src/lib.rs
Comment (aphelionz) [inline src/lib.rs]: What about changing this to `#![warn(missing_docs)]` or something that doesn't break CI but also doesn't just eschew code-level docs altogether?

I can see that this PR is substantial and I appreciate that not every single function and trait tediously needs a comment like `// Decodes CBOR` but I'd rather err on the side of over-documenting than under-documenting.

**Review by aphelionz — CHANGES_REQUESTED** [review]
Overall looks really good to me. Thanks for writing the new tests and for doing the legwork of changing the semantics of every encoding type. Just a few minor things and I'm happy to approve, but want @vmx at least to weigh in as well.

File: core/src/mem.rs
Comment (dvc94ch) [reply core/src/mem.rs]: I think it's useful as a reference implementation. Any store implementation should be "equivalent". If you want to operate on ipld in memory it does make sense to have such an implementation. But they are good points.

Comment (aphelionz) [reply core/src/mem.rs]: Agree it's a great reference

Comment (dvc94ch) [reply core/src/mem.rs]: Well it would be if it were correct, but I think I'll leave that for a follow up.

File: core/src/mem.rs
Comment (dvc94ch) [reply core/src/mem.rs]: will be fixed in a follow up

**Comment by dvc94ch** [comment]
@vmx any comments? I know you want to make it generic over the `Code`, can that be done in a follow up?

File: core/src/codec.rs
Comment (vmx) [inline core/src/codec.rs]: I'd remove the version. This is part of a CID and not of a Codec.

File: core/src/ipld.rs
Comment (vmx) [inline core/src/ipld.rs]: I think vectors are way more common than Boxed slices. So I'm not sure about this change.

File: core/src/ipld.rs
Comment (vmx) [inline core/src/ipld.rs]: This sounds very similar what is described as `Node` in the [IPLD Library Design docs](https://github.com/ipld/specs/blob/4eb7b233f38864262c24cc91203098f483f21432/design/libraries/nodes-and-kinds.md#node). So perhaps it would make sense to also introduce the concept of a `Node`.

It would also be good to have some more docs about what differentiates this type from `Ipld`.

File: core/src/store.rs
Comment (vmx) [inline core/src/store.rs]: I'm not sure if the storage part should really be part of `core`. But as it currently already is, it is outside of scope for this PR and we should discuss it on a separate issue.

File: src/block.rs
Comment (vmx) [inline src/block.rs:2]: Here is where my proposal of the Block API (https://github.com/ipfs-rust/rust-ipld/issues/14) comes in. I've created a small demo application which used the proposed API (some details changed, but the idea is the same). The full implementation of that API can be found here: https://github.com/vmx/rust-dagbuilder/blob/f7f0c00b96074f6aa6a5aeef0b327b35d73dea72/src/block.rs

More information is available on the [Block API issue](https://github.com/ipfs-rust/rust-ipld/issues/14#issuecomment-624324812).

I hope that we can use my proposed API (of course after further discussions) instead.

My code currently uses a modified libipld, tomorrow I'll try to base it on this PR, then I can also say more about the encoder/decoder changes.

**Review by vmx — COMMENTED** [review]
I think we should have a discussion about whether we want to use `Vec<u8>` or `Box<[u8]>` for a lot of things. I need to do more research on it as I currently don't see benefits of using a Boxed slice instead of a vector. So if anyone has pros/cons for either of those, it would be really helpful to post them here.

File: core/src/codec.rs
Comment (dvc94ch) [reply core/src/codec.rs]: good point

File: core/src/ipld.rs
Comment (dvc94ch) [reply core/src/ipld.rs]: maybe true, but an immutable opaque blob of bytes should be a `Box<[u8]>` not a `Vec<u8>`.

I mentioned in another PR that `Encode` can only be implemented for `Vec<T>` and not for `Vec<T>` and `Vec<u8>`, and same for `Decode`.

File: core/src/ipld.rs
Comment (dvc94ch) [reply core/src/ipld.rs]: This type is only used for constructing `TypeError`.

File: src/block.rs
Comment (dvc94ch) [reply src/block.rs:2]: One thing I don't like is this:
```rust
Box::<dyn Codec>::from(self.codec).decode(&raw)
```
There needs to be a trait Code:
```rust
trait Code {
    fn encode(&self) {
        match self {
            Self::Cbor => Cbor::encode(),
        }
    }
    fn decode(&self)
}

impl Code for Codec {}
```
same in multihash there needs to be a trait Code:
```rust
trait Code {
    fn digest()
}
```

Comment (dvc94ch) [reply src/block.rs:2]: encode and decode should return a reference, the caller can clone if ownership is required.

the block api also doesn't support encoding structs directly, Ipld is required as an intermediate representation.

I don't see necessarily that the two api's are conflicting, they have different use-cases.

File: dag-cbor/src/encode.rs
Comment (vmx) [inline dag-cbor/src/encode.rs]: There is also the identity hash, which is just the data itself. There 128 byte wouldn't be enough.

Comment (dvc94ch) [reply dag-cbor/src/encode.rs]: the identity hash doesn't have a size restriction?

Comment (dvc94ch) [reply dag-cbor/src/encode.rs]: I believe that would be a bug. The reason for the identity hash is when the data is smaller than a cryptographic hash function. So it can not be longer than the longest hash function we have which would be 512bits.

https://github.com/multiformats/multihash/issues/13

File: src/block.rs
Comment (vmx) [reply src/block.rs:2]: > the block api also doesn't support encoding structs directly, Ipld is required as an intermediate representation.

That's something I haven't thought about yet. But it's a good point. So far I've seen the `Ipld` enum as the central piece everything should center around. But perhaps we can get away with using native types. I need to thing about that.

Also perhaps the block api can be extended to also support native types.

File: core/src/store.rs
Comment (dvc94ch) [reply core/src/store.rs]: do you want to open an issue for it?

**Comment by vmx** [comment]
I just want to mention that I had a quick look now and things look good. I'll have a proper review tomorrow, but I wouldn't expect anything major.

File: Cargo.toml
Comment (aphelionz) [inline Cargo.toml]: Another reminder to remove this before merging

File: core/src/convert.rs
Comment (aphelionz) [inline core/src/convert.rs:43]: Looks good. Nice to see both `Box` and `Vec` supported here

File: dag-json/src/codec.rs
Comment (aphelionz) [inline dag-json/src/codec.rs]: Why inline only on these? or did I miss others?

**Review by aphelionz — CHANGES_REQUESTED** [review]
Overall looks good, just a few minor things. Also, at some point we'll need a proper README.md with usage and some form of API documentation but that can obviously be left for later.

File: dag-json/src/codec.rs
Comment (dvc94ch) [reply dag-json/src/codec.rs]: We should probably remove all #[inline] as they increase compile times and recommend that users do lto in release builds.

https://internals.rust-lang.org/t/when-should-i-use-inline/598/2

Comment (dvc94ch) [reply dag-json/src/codec.rs]: Our encode/decode is generic over a Reader/Writer so it may even do inlining without the attribute

File: core/src/store.rs
Comment (vmx) [reply core/src/store.rs]: Done: https://github.com/ipfs-rust/rust-ipld/issues/38

Comment (dvc94ch) [reply core/src/store.rs]: I moved it out of core, is that good enough?

Comment (vmx) [reply core/src/store.rs]: Yes, sorry, that's good enough. I thought misread `src` for being `core`.

**Comment by dvc94ch** [comment]
One thing left to do is move the errors out of core

**Comment by koivunej** [comment]
If I understand correctly that the changes can be summarized as:

 1. unification of encoding and decoding through respective per codec traits from `WriteCbor` and plain fns for json, pb codecs
 2. internal `ReadCbor` renamed to `TryReadCbor`
 3. inside codecs the errors now are named `Error`
 4. a lot of renames following the (1), (2), (3)
 5. `#[inline]` cleanup
 6. removals in the top level crate

None of these seem major changes in functionality but mostly "streamlining".

Did I miss something?

**Comment by dvc94ch** [comment]
You can call it what you want. But together with the ipld-block-builder and ipfs-embed it provides a complete ipfs implementation.

**Comment by koivunej** [comment]
> You can call it what you want.

I summarized the changes I went through in order for us to make sure I didn't miss anything. I feel like you read me criticizing the work which I did not intend to do.

**Comment by dvc94ch** [comment]
Maintenance is work. The new Codec trait is much more flexible and the status of cbor as a special codec is removed. The only thing special now is that it supports native types, but someone can contribute that to dag-json or dag-pb if they like.

**Review by vmx — APPROVED** [review]
Thanks for the effort! Things look good now.

Please wait for one more approval from the Rust IPFS folks, before merging this.

**Review by koivunej — APPROVED** [review]

**Review by aphelionz — APPROVED** [review]

**Comment by aphelionz** [comment]
Squash one more time and then merge away!

**Pull request merged**
